### PR TITLE
multiVI with weights

### PR DESCRIPTION
**Pull request by marianogabitto**
FIRST COMMIT WITH WEIGHT SUPPORT.

### CODE REVIEW

**Review by marianogabitto — COMMENTED** [review]
No comment

File: scvi/module/_utils.py [hunk 1]
```diff
@@ -41,3 +41,16 @@ def batch(batch_size, label):
 
     batch_size = x.size(0)
     return torch.cat([batch(batch_size, i) for i in range(y_dim)])
+
+
+def masked_softmax(weights, mask, dim=-1, eps=1e-30):
```
Comment (adamgayoso) [inline scvi/module/_utils.py:46]: can you add a proper docstring and typing here?

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: scvi/module/_utils.py (hunk 1 above)
Comment (marianogabitto) [reply scvi/module/_utils.py:46]: Done, first line is summary.

File: scvi/module/_multivae.py [hunk 2]
```diff
@@ -523,37 +601,67 @@ def get_reconstruction_loss_expression(self, x, px_rate, px_r, px_dropout):
...
-    def _mix_modalities(x_paired, x_expr, x_acc, mask_expr, mask_acc):
-        """
-        Mixes modality-specific vectors according to the modality masks.
-
-        in positions where both `mask_expr` and `mask_acc` are True (corresponding to cell
-        for which both expression and accessibility data is available), values from `x_paired`
-        will be used. If only `mask_expr` is True, use values from `x_expr`, and if only
-        `mask_acc` is True, use values from `x_acc`.
+    def _compute_mod_penalty(self, mod_params1, mod_params2, mask1, mask2):
+        """Weighted mean of Xs while masking values that originate from non-measured modalities
```
Comment (adamgayoso) [inline scvi/module/_multivae.py:612]:
```suggestion
        """
        Weighted mean of Xs while masking values that originate from non-measured modalities.
```

File: scvi/module/_multivae.py [hunk 3]
```diff
@@ -523,37 +601,67 @@ def get_reconstruction_loss_expression(self, x, px_rate, px_r, px_dropout):
...
+            torch.logical_and(mask1, mask2),
+            pair_penalty.T,
+            torch.zeros_like(pair_penalty).T,
+        ).sum(dim=0)
+
+
+@auto_move_data
+def mix_modalities(Xs, masks, weights, weight_transform: callable = None):
+    """Compute the weighted mean of the Xs while masking values that originate
+    from modalities that aren't measured.
```
Comment (adamgayoso) [inline scvi/module/_multivae.py:641]:
```suggestion
    """
    Compute the weighted mean of the Xs while masking unmeasured modality values.
```

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: scvi/module/_multivae.py [hunk 4]
```diff
@@ -55,13 +57,27 @@ class MULTIVAE(BaseModuleClass):
         Number of input regions.
     n_input_genes
         Number of input genes.
+    modality_weights
+        "equal": Equal weight in each modality
+        "universal": Learn weights across modalities w_m.
+         "cell": Learn weights across modalities and cells. w_{m,c}
```
Comment (adamgayoso) [inline scvi/module/_multivae.py:63]:
```suggestion
        * ``"equal"``: Equal weight in each modality
        * ``"universal"``: Learn weights across modalities w_m.
        * ``"cell"``: Learn weights across modalities and cells. w_{m,c}
```

File: scvi/module/_multivae.py [hunk 5]
```diff
@@ -55,13 +57,27 @@ class MULTIVAE(BaseModuleClass):
...
+    modality_penalty
+        "Jeffreys": Jeffreys penalty to align modalities
+         "MMD": MMD penalty to align modalities
+          "None": No penalty
```
Comment (adamgayoso) [inline scvi/module/_multivae.py:67]:
```suggestion
        * ``"Jeffreys"``: Jeffreys penalty to align modalities
        * ``"MMD"``: MMD penalty to align modalities
        * ``"None"``: No penalty
```

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: scvi/model/_multivi.py [hunk 6]
```diff
@@ -872,9 +895,22 @@ def setup_anndata(
...
         cls.register_manager(adata_manager)
+
+    def _check_adata_modalityweights(self, adata):
+        """
+        Checks if adata is None and weights are per cell.
+
+        :param adata: anndata object
+        :return:
+        """
+        if (adata is not None) and (self.modality_weights == "cell"):
```
Comment (adamgayoso) [inline scvi/model/_multivi.py:913]:
```suggestion
        if (adata is not None) and (self.module.modality_weights == "cell"):
```

File: scvi/module/_multivae.py [hunk 7]
```diff
@@ -55,13 +57,27 @@ class MULTIVAE(BaseModuleClass):
...
     n_input_genes
         Number of input genes.
+    modality_weights
+        * ``"equal"``: Equal weight in each modality
+        * ``"universal"``: Learn weights across modalities w_m.
+        * ``"cell"``: Learn weights across modalities and cells. w_{m,c}
+    modality_penalty
+        * ``"Jeffreys"``: Jeffreys penalty to align modalities
+        * ``"MMD"``: MMD penalty to align modalities
+        * ``"None"``: No penalty
```
Comment (adamgayoso) [inline scvi/module/_multivae.py:67]: Can you add the proper docstrings (these new parameters) as well in MULTIVI class?

Comment (adamgayoso) [reply scvi/module/_multivae.py:67]: Also line 83 of _multivi.py should say MULTIVAE not PEAKVAE

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: scvi/module/_multivae.py (hunk 7 above)
Comment (marianogabitto) [reply scvi/module/_multivae.py]: MULTIVAE, SOLVED.

ADDED DOCSTRING

**Pull request merged**